File: src/engine/Camera/Camera.jl
module CameraModule
    using ..JulGame
    using .Math

    export Camera, pixels_per_world_unit

    mutable struct Camera
        id::String
        name::String
		backgroundColor::NTuple{4, Int}
        offset::Vector2f
        position::Vector3f
        size::Vector2
        zoom::Float64
        yaw::Float64
        pitch::Float64
        target::Union{
            Ptr{Nothing}, 
            JulGame.TransformModule.Transform
            }
        windowPos::Vector2

        function Camera(size::Vector2, initialPosition::Vector3f, offset::Vector2f, target)
            this = new()
            
            this.id = JulGame.generate_uuid()
            this.name = "Camera"
            this.backgroundColor = (0,0,0, 255)
            this.size = size
            this.position = initialPosition
            this.offset = Vector2f(offset.x, offset.y)
            this.target = target
            this.windowPos = Vector2(0,0)
            this.zoom = 1.0
            this.yaw = 0.0
            this.pitch = 0.0

            return this
        end
    end

    """Pixels per world unit for 2D rendering (SCALE_UNITS × camera zoom)."""
    @inline function pixels_per_world_unit(camera::Union{Nothing, Camera})
        camera === nothing && return JulGame.SCALE_UNITS
        return JulGame.SCALE_UNITS * camera.zoom
    end

    function update(this::Camera)
        if !JulGame.IS_EDITOR && JulGame.WindowManagerModule.get_logical_size().x != this.size.x || JulGame.WindowManagerModule.get_logical_size().y != this.size.y
            JulGame.WindowManagerModule.set_logical_size(this.size.x, this.size.y)
            @debug "Logical size changed to $(this.size)"
        end

        if JGStaticModule.LIB_AVAILABLE
            ccall(
                (:static_update_camera, JGStaticModule.LIB_PATH),
                Cvoid,
                (Ptr{Cvoid}, Ptr{Cvoid}, Int64),
                pointer_from_objref(this),
                Ptr{Cvoid}(JulGame.Renderer),
                reinterpret(Int64, Float64(JulGame.SCALE_UNITS)),
            )
        else
            SDL2.SDL_SetRenderDrawBlendMode(JulGame.Renderer::Ptr{SDL2.SDL_Renderer}, SDL2.SDL_BLENDMODE_BLEND)
            rgba = (r = Ref(UInt8(0)), g = Ref(UInt8(0)), b = Ref(UInt8(0)), a = Ref(UInt8(255)))
            SDL2.SDL_GetRenderDrawColor(JulGame.Renderer::Ptr{SDL2.SDL_Renderer}, rgba.r, rgba.g, rgba.b, rgba.a)
            SDL2.SDL_SetRenderDrawColor(Renderer, this.backgroundColor[1], this.backgroundColor[2], this.backgroundColor[3], this.backgroundColor[4]);
            SDL2.SDL_RenderFillRectF(Renderer, Ref(SDL2.SDL_FRect(this.windowPos.x, this.windowPos.y, this.size.x, this.size.y)))
            SDL2.SDL_SetRenderDrawColor(JulGame.Renderer::Ptr{SDL2.SDL_Renderer}, rgba.r[], rgba.g[], rgba.b[], rgba.a[]);
            
            center_pixels = Vector2f(this.size.x / 2, this.size.y / 2)
            center_world = center_pixels / pixels_per_world_unit(this)

            if this.target !== nothing && this.target !== C_NULL
                targetPos::Vector3f = this.target.position
                targetScale::Vector2f = this.target.scale
                this.position = Vector3f(targetPos.x - center_world.x + 0.5 * targetScale.x + this.offset.x,
                                         targetPos.y - center_world.y + 0.5 * targetScale.y + this.offset.y, 
                                         targetPos.z)
                return
            end
        end
    end

    # making set property observable
    function Base.setproperty!(this::Camera, s::Symbol, x)
        @debug("setting camera property $(s) to: $(x)")
        try
            setfield!(this, s, x)
        catch e
            println(e)
        end
    end
end

File: ts/_generated/src/engine/Camera/Camera.ts
export {}

    // using ..JulGame
    // using .Math

    

    class Camera {
        id: string
        name: string
		backgroundColor: [number, number, number, number]
        offset: Vector2f
        position: Vector3f
        size: Vector2
        zoom: number
        yaw: number
        pitch: number
        target: null | ITransform
        windowPos: Vector2

        constructor(size: Vector2, initialPosition: Vector3f, offset: Vector2f, target) {
            
            
            this.id = (globalThis as any).JulGame.generate_uuid()
            this.name = "Camera"
            this.backgroundColor = [0,0,0, 255]
            this.size = size
            this.position = initialPosition
            this.offset = {x: offset.x, y: offset.y}
            this.target = target
            this.windowPos = {x: 0, y: 0}
            this.zoom = 1.0
            this.yaw = 0.0
            this.pitch = 0.0

        }
    }

    /*Pixels per world unit for 2D rendering (SCALE_UNITS × camera zoom).*/
    function pixels_per_world_unit(camera: null | Camera) {
        if (camera === null) { return (globalThis as any).JulGame.SCALE_UNITS }
        return (globalThis as any).JulGame.SCALE_UNITS * camera.zoom
    }

    /*
        apply_zoom_to_center(camera, new_zoom: number)

    Set `camera.zoom` to `new_zoom` and shift `camera.position` (x, y) so the world point
    that was at the viewport center stays there. `camera.offset` is not modified.
    */
    function apply_zoom_to_center(camera: Camera, new_zoom: number) {
        let S_old = pixels_per_world_unit(camera)
        let S_new = (globalThis as any).JulGame.SCALE_UNITS * new_zoom
        let half_w = Number(camera.size.x) / 2
        let half_h = Number(camera.size.y) / 2
        let inv_delta = (1 / (S_old)) - (1 / (S_new))
        let dx = half_w * inv_delta
        let dy = half_h * inv_delta
        camera.position = {x: camera.position.x + dx, y: camera.position.y + dy, z: camera.position.z}
        camera.zoom = new_zoom
        return camera
    }

    function update(self: Camera, newPosition: null | Vector3f = null) {
        if ((globalThis as any).JulGame.IS_EDITOR && (globalThis as any).JulGame.WindowManagerModule.get_logical_size() != self.size) {
            (globalThis as any).JulGame.WindowManagerModule.set_logical_size(self.size.x, self.size.y)
            console.debug(`Logical size changed to ${self.size}`)
        }
(globalThis as any).JulGameSdl.glue_SDL_SetRenderDrawBlendMode_BLEND()
        let rgba = { r: 0, g: 0, b: 0, a: 255 }
        (globalThis as any).JulGameSdl.glue_SDL_GetRenderDrawColor((globalThis as any).JulGame.Renderer, rgba.r, rgba.g, rgba.b, rgba.a);
        (globalThis as any).JulGameSdl.glue_SDL_SetRenderDrawColor(self.backgroundColor[0], self.backgroundColor[1], self.backgroundColor[2], self.backgroundColor[3]);
        (globalThis as any).JulGameSdl.glue_SDL_RenderFillRectF((globalThis as any).JulGameSdl.glue_SDL_FRect(self.windowPos.x, self.windowPos.y, self.size.x, self.size.y));
        (globalThis as any).JulGameSdl.glue_SDL_SetRenderDrawColor((globalThis as any).JulGame.Renderer, rgba.r, rgba.g, rgba.b, rgba.a);
        
        let center_pixels = {x: self.size.x / 2, y: self.size.y / 2}
        let center_world = {x: center_pixels.x / pixels_per_world_unit(self), y: center_pixels.y / pixels_per_world_unit(self)}

        if (self.target !== null && self.target !== null && newPosition === null) {
            let targetPos = self.target.position
            let targetScale = self.target.scale
            self.position = {x: targetPos.x - center_world.x + 0.5 * targetScale.x + self.offset.x, y: targetPos.y - center_world.y + 0.5 * targetScale.y + self.offset.y, z: targetPos.z}
            return
        }

        if (newPosition !== null) {
            self.position = newPosition
        }
    }

    // making set property observable

